src/charm.py

```diff
@@ -48,66 +71,68 @@ def _get_pebble_config(self, event):
         }
         return pebble_config
 
-    def _on_config_changed(self, event):
-        """Handle config-changed event."""
+    def _on_config_changed(self, event: ConfigChangedEvent) -> None:
+        """Handle config-changed event.
 
-        if not self._is_valid_config():
-            # Charm will be in blocked status.
+        Args:
+            event: The information about the event.
+        """
+        # Check for container connectivity
+        if not self.unit.get_container(self.service_name).can_connect():
+            event.defer()
             return
 
-        pebble_config = self._get_pebble_config(event)
-        container = self.unit.get_container(self.service_name)
+        # Check whether configuration is valid
+        config_valid, message = self._is_valid_config()
+        if not config_valid:
+            logger.info(message)
+            self.unit.status = BlockedStatus(message)
+            return
 
+        # Add any newly required or update any changed services
+        pebble_config = self._get_pebble_config()
+        container = self.unit.get_container(self.service_name)
         services = container.get_plan().to_dict().get("services", {})
         if services != pebble_config["services"]:
-            logger.debug("About to add_layer with pebble_config:\n{}".format(yaml.dump(pebble_config)))
+            logger.debug("About to add_layer with pebble_config:\n%s", yaml.dump(pebble_config))
             container.add_layer(self.service_name, pebble_config, combine=True)
-
-            self._restart_service(self.service_name, container)
+            container.restart(self.service_name)
```

Comment:
Won't `pebble.replan_services` achieve the same result without the need to check if there are changes?

Reply:
Thank you, fixed